src/core/dispatcher.rs

```diff
@@ -1,121 +1,133 @@
-    ///
-    /// # Arguments
-    ///
-    /// * handler - A handler to add
+    /// * handler - Handler to add
     ///
     /// Handlers will be dispatched in the same order as they are added
     pub fn add_handler<H, I>(&mut self, handler: H) -> &mut Self
     where
         H: Handler<I> + Sync + Clone + 'static,
-        I: TryFromInput + 'static,
+        I: TryFromInput + Sync + 'static,
         <H::Future as Future>::Output: Into<HandlerResult>,
     {
-        self.handlers.push(ConvertHandler::boxed(handler));
+        self.handlers.push(ConvertInputHandler::boxed(handler));
         self
     }
 
-    /// Sets a handler to be executed when an error has occurred
-    ///
-    /// # Arguments
-    ///
-    /// * handler - A handler to set
-    ///
-    /// Error handler will be called if one of update handlers returned
-    /// [HandlerResult::Error](enum.HandlerResult.html)
-    ///
-    /// If this method is not called,
-    /// [LoggingErrorHandler](struct.LoggingErrorHandler.html)
-    /// will be used as default handler.
-    pub fn set_error_handler<H>(&mut self, handler: H) -> &mut Self
-    where
-        H: ErrorHandler + Sync + 'static,
-    {
-        self.error_handler = Arc::new(ConvertErrorHandler::boxed(handler));
-        self
+    /// Creates a new dispatcher
+    pub fn build(self) -> Dispatcher {
+        Dispatcher {
+            handlers: Arc::new(self.handlers),
+        }
     }
+}
 
-    fn dispatch(&self, update: Update) -> impl Future<Output = ()> {
-        let input = HandlerInput {
-            update,
-            context: self.context.clone(),
-        };
-        let futures: Vec<_> = self.handlers.iter().map(|h| h.handle(input.clone())).collect();
-        let error_handler = self.error_handler.clone();
+/// Updates dispatcher
+#[derive(Clone)]
+pub struct Dispatcher {
+    handlers: Arc<Vec<Box<dyn InputHandler + Sync>>>,
+}
+
+impl Dispatcher {
+    fn dispatch(&self, input: HandlerInput) -> impl Future<Output = HandlerResult> {
+        let handlers = self.handlers.clone();
         async move {
-            for future in futures {
-                let result = future.await;
-                match result {
-                    HandlerResult::Continue => {}
-                    HandlerResult::Stop => break,
-                    HandlerResult::Error(err) => {
-                        error_handler.handle(err).await;
-                        break;
-                    }
+            for handler in handlers.iter() {
+                let type_name = handler.get_type_name();
+                log::debug!("Running '{}' handler...", type_name);
+                let result = handler.handle(input.clone()).await;
+                if matches!(result, HandlerResult::Stop | HandlerResult::Error(_)) {
+                    log::debug!("'{}' handler returned {:?}, loop stopped", type_name, result);
+                    return result;
                 }
             }
+            HandlerResult::Stop
```

Comment:
Why Stop?

Reply:
Because we always return Stop by default (excluding predicates and other service handlers).

Reply:
Dispatcher is just a list of handlers, so what's point if it always stops? We can't do this then:
```rust
let pre = Dispatcher::new();
a.add_handler(my_handler1);
a.add_handler(my_handler2);

let post = Dispatcher::new();
b.add_handler(my_handler3);
b.add_handler(my_handler4);

let together = Dispatcher::new();
together.add_handler(pre);
together.add_handler(something_between);
together.add_handler(post);
```

Reply:
Okay, that is right